File: Example/PayTPV Objc Example/AlertBuilder.m
//
//  AlertBuilder.m
//  PAYTPV Objc Example
//
//  Created by Mihail Cristian Dumitru on 8/21/17.
//  Copyright © 2017 PAYTPV. All rights reserved.
//

#import "AlertBuilder.h"

@implementation AlertBuilder

- (instancetype)init {
    self = [super init];
    if (self) {
        _alertController = [UIAlertController alertControllerWithTitle:nil message:nil preferredStyle:UIAlertControllerStyleAlert];
    }
    return self;
}

- (instancetype)setTitle:(NSString *)title {
    self.alertController.title = title;
    return self;
}

- (instancetype)setMessage:(NSString *)message {
    self.alertController.message = message;
    return self;
}

- (instancetype)addActionWithTitle:(NSString *)title
                             style:(UIAlertActionStyle)style
                           handler:(void (^ __nullable)(UIAlertAction * action))handler {
    [self.alertController addAction:[UIAlertAction actionWithTitle:title
                                                             style:style
                                                           handler:handler]];
    return self;
}

- (instancetype)addLoadingIndicator {
    UIActivityIndicatorView *indicator = [UIActivityIndicatorView new];
    indicator.translatesAutoresizingMaskIntoConstraints = false;
    [self.alertController.view addSubview:indicator];

    NSDictionary *views = @{
                            @"pending": self.alertController.view,
                            @"indicator": indicator,
                            };
    NSArray *constraints = [NSLayoutConstraint constraintsWithVisualFormat:@"V:[indicator]-(-50)-|"
                                                                   options:0
                                                                   metrics:nil
                                                                     views:views];
    constraints = [constraints arrayByAddingObjectsFromArray:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|[indicator]|"
                                                                                                     options:0
                                                                                                     metrics:nil
                                                                                                       views:views]];
    [self.alertController.view addConstraints:constraints];

    [indicator setUserInteractionEnabled:false];
    [indicator startAnimating];

    return self;
}

- (instancetype)presentInController:(UIViewController *)sender
                           animated:(BOOL)animated
                         completion:(void (^ __nullable)(void))completion {
    [sender presentViewController:self.alertController animated:animated completion:completion];

    return self;
}

- (void)dismissAnimated:(BOOL)animated
             completion:(void (^ __nullable)(void))completion {
    [self.alertController dismissViewControllerAnimated:animated completion:completion];
}

@end

File: Example/PayTPV Swift Example/AlertBuilder.swift
//
//  AlertBuilder.swift
//  PAYTPV Swift Example
//
//  Created by Mihail Cristian Dumitru on 8/18/17.
//  Copyright © 2017 PAYTPV. All rights reserved.
//

import UIKit

class AlertBuilder {

    // MARK: - Properties

    var alertController = UIAlertController(title: nil, message: nil, preferredStyle: .alert)

    // MARK: - Helpers

    func set(title: String?) -> Self {
        self.alertController.title = title
        return self
    }

    func set(message: String?) -> Self {
        self.alertController.message = message
        return self
    }

    func add(actionWithTitle title: String?, style: UIAlertActionStyle = .default, handler: ((UIAlertAction) -> Void)? = nil) -> Self {
        self.alertController.addAction(
            UIAlertAction(
                title: title,
                style: style,
                handler: handler
            )
        )
        return self
    }

    func addLoadingIndicator() -> Self {
        let indicator = UIActivityIndicatorView()
        indicator.translatesAutoresizingMaskIntoConstraints = false
        alertController.view.addSubview(indicator)

        let views = ["pending" : alertController.view, "indicator" : indicator]
        var constraints = NSLayoutConstraint.constraints(withVisualFormat: "V:[indicator]-(-50)-|", options: [], metrics: nil, views: views)
        constraints += NSLayoutConstraint.constraints(withVisualFormat: "H:|[indicator]|", options: [], metrics: nil, views: views)
        alertController.view.addConstraints(constraints)

        indicator.isUserInteractionEnabled = false
        indicator.startAnimating()

        return self
    }

    func present(sender: UIViewController, animated: Bool = true, completion: (() -> Void)? = nil) -> Self {
        sender.present(self.alertController, animated: animated, completion: completion)

        return self
    }

    func dismiss(animated: Bool = true, completion:(() -> Void)? = nil) {
        self.alertController.dismiss(animated: animated, completion: completion)
    }

}
